cloudify_agent/operations.py

```diff
@@ -165,3 +182,98 @@ def _save_daemon(daemon):
+        return new_agent_name
+    else:
+        return '{0}_{1}'.format(old_agent_name, suffix)
+
+
+def create_new_agent_dict(old_agent, suffix=None):
+    if not suffix:
+        suffix = str(uuid.uuid4())
+    new_agent = copy.deepcopy(old_agent)
+    fields_to_delete = ['name', 'queue', 'workdir', 'agent_dir', 'envdir',
+                        'manager_ip', 'package_url', 'source_url'
+                        'manager_port', 'broker_url', 'broker_port',
+                        'broker_ip', 'key', 'version']
+    for field in fields_to_delete:
+        if field in new_agent:
+            del(new_agent[field])
+    new_agent['name'] = _generate_new_agent_name(old_agent['name'], suffix)
+    new_agent['remote_execution'] = True
+    configuration.reinstallation_attributes(new_agent)
+    new_agent['manager_file_server_url'] = get_manager_file_server_url()
+    # Assuming that if there is no version info in the agent then
+    # this agent was installed by current manager.
+    new_agent['old_agent_version'] = old_agent.get('version', _VERSION)
+    return new_agent
+
+
+def create_agent_from_old_agent(install_agent_timeout=300):
+    if 'cloudify_agent' not in ctx.instance.runtime_properties:
+        raise NonRecoverableError(
+            'cloudify_agent key not available in runtime_properties')
+    old_agent = ctx.instance.runtime_properties['cloudify_agent']
+    new_agent = create_new_agent_dict(old_agent)
+    # We retrieve broker url from old agent in order to support
+    # cases when old agent is not connected to current rabbit server.
+    broker_url = _get_broker_url(ctx, old_agent)
+    env_broker_url = os.environ.get('CELERY_BROKER_URL')
+    os.environ['CELERY_BROKER_URL'] = broker_url
+    try:
+        celery_client = celery.Celery(broker=broker_url, backend=broker_url)
+        if new_agent['old_agent_version'] != '3.2':
+            celery_client.conf.update(
+                CELERY_TASK_RESULT_EXPIRES=defaults.CELERY_TASK_RESULT_EXPIRES
+            )
+        script_format = '{0}/cloudify/install_agent.py'
+        script_url = script_format.format(get_manager_file_server_url())
+        result = celery_client.send_task(
+            'script_runner.tasks.run',
+            args=[script_url],
+            kwargs={'cloudify_agent': new_agent},
+            queue=old_agent['queue']
+        )
+        returned_agent = result.get(timeout=install_agent_timeout)
+    finally:
+        if env_broker_url is None:
+            del(os.environ['CELERY_BROKER_URL'])
+        else:
+            os.environ['CELERY_BROKER_URL'] = env_broker_url
+    # Make sure that new celery agent was started:
+    agent_status = app.control.inspect(destination=[
+        'celery@{0}'.format(returned_agent['name'])])
+    if not agent_status.active():
+        raise NonRecoverableError('Could not start agent.')
+    # Setting old_cloudify_agent in order to uninstall it later.
+    ctx.instance.runtime_properties['old_cloudify_agent'] = old_agent
```

Comment:
Uninstall is not available yet.
I am working on this but it will not be added in this iteration.